From b5261e3a4defc686df24c414052a7ceba1d8060f Mon Sep 17 00:00:00 2001
From: Zaikuo Wang <zaikuo.wang@intel.com>
Date: Tue, 5 May 2015 17:00:57 +0800
Subject: [PATCH 383/462] ov2722: retry startup when powerup fail

On CHT, ov2722 power on failed when run stress test;
retry to powerUp at most 5 times if startUp failed at first time

Change-Id: I5c94a0026a32f1990b8cb1db6aec14d930acc53e
Tracked-On:https://jira01.devtools.intel.com/browse/IMINAN-30744
Signed-off-by: Zaikuo Wang <zaikuo.wang@intel.com>
Reviewed-on: https://icggerrit.ir.intel.com/54304
Reviewed-by: Shuguang Gong <shuguang.gong@intel.com>
Validation-Android: Wei Tao <wei.tao@intel.com>
Approver: Shuguang Gong <shuguang.gong@intel.com>
Build: IMIN (SC)
Integrator: WeiX Meng <weix.meng@intel.com>
Build: WeiX Meng <weix.meng@intel.com>
Reviewed-by: WeiX Meng <weix.meng@intel.com>
Maintainer: WeiX Meng <weix.meng@intel.com>
---
 .../camera/drivers/media/i2c/ov2722.c              | 26 ++++++++++++++++++++--
 .../camera/drivers/media/i2c/ov2722.h              |  2 ++
 2 files changed, 26 insertions(+), 2 deletions(-)

diff --git a/drivers/external_drivers/camera/drivers/media/i2c/ov2722.c b/drivers/external_drivers/camera/drivers/media/i2c/ov2722.c
index 9cf1e4c..598dbcd 100644
--- a/drivers/external_drivers/camera/drivers/media/i2c/ov2722.c
+++ b/drivers/external_drivers/camera/drivers/media/i2c/ov2722.c
@@ -1002,8 +1002,30 @@ static int ov2722_s_mbus_fmt(struct v4l2_subdev *sd,
 
 	ret = startup(sd);
 	if (ret) {
-		dev_err(&client->dev, "ov2722 startup err\n");
-		goto err;
+		int i = 0;
+		dev_err(&client->dev, "ov2722 startup err, retry to power up\n");
+		for (i = 0; i < OV2722_POWER_UP_RETRY_NUM; i++) {
+			dev_err(&client->dev,
+				"ov2722 retry to power up %d/%d times, result: ",
+				i+1, OV2722_POWER_UP_RETRY_NUM);
+			power_down(sd);
+			ret = power_up(sd);
+			if (ret) {
+				dev_err(&client->dev, "power up failed, continue\n");
+				continue;
+			}
+			ret = startup(sd);
+			if (ret) {
+				dev_err(&client->dev, " startup FAILED!\n");
+			} else {
+				dev_err(&client->dev, " startup SUCCESS!\n");
+				break;
+			}
+		}
+		if (ret) {
+			dev_err(&client->dev, "ov2722 startup err\n");
+			goto err;
+		}
 	}
 
 	ret = ov2722_get_intg_factor(client, ov2722_info,
diff --git a/drivers/external_drivers/camera/drivers/media/i2c/ov2722.h b/drivers/external_drivers/camera/drivers/media/i2c/ov2722.h
index 41e0064..7621608 100644
--- a/drivers/external_drivers/camera/drivers/media/i2c/ov2722.h
+++ b/drivers/external_drivers/camera/drivers/media/i2c/ov2722.h
@@ -40,6 +40,8 @@
 
 #define OV2722_NAME		"ov2722"
 
+#define OV2722_POWER_UP_RETRY_NUM 5
+
 /* Defines for register writes and register array processing */
 #define I2C_MSG_LENGTH		0x2
 #define I2C_RETRY_COUNT		5
-- 
1.9.1

